### DISCUSSION TITLE
HomeSpan 2.0.0-rc.1

### DISCUSSION THREAD

**Discussion by HomeSpan** [HomeSpan/HomeSpan#923] [General]
### Re-factored for compatibility with Arduino-ESP32 Board Manager version 3

* requires Arduino-ESP32 Board Manager 3.0.2 or later
  * Espressif did not include the Sodium cryptography library in versions 3.0.0 and 3.0.1, which is required by HomeSpan
* all user-sketches written under HomeSpan 1.9.1 should work without modification under HomeSpan 2.0.0, with the exception of sketches in which the user directly accessed IDF-4 functionality, which may or may not be compatible with IDF-5 
 * supports use of the ESP32-C6 chip
   * though the Arduino-ESP32 Board Manager also supports the ESP-H2, HomeSpan does not since the H2 does not have WiFi capabilities, as currently required
  
### New API Functions

* **Complete re-write of the HomeSpan Pixel and RFControl libraries**

  * you can now instantiate both **Pixel** and **RFControl** objects in the same sketch (previously these classes were incompatible with eachother and could not be used in the same sketch)

 * **Expanded **Pixel** class functionality**
      * adds support for 5-color RGBWC (red, green, blue, warm-white, and cool-white) LEDs
      * adds new method `Pixel::Color::WC()` to support setting warm-white / cool-white color values
      * adds new methods `Pixel::Color::CCT()` and `Pixel::setTemperatures()` to automatically set the warm-white / cool-white colors of a Pixel device based on whatever correlated color temperature (CCT) the user specifies (e.g. 3000K)
      * adds new constructor `Pixel(int pin, const char *pixelType)` that provides the ability to both select the colors and specify their transmission order (*pixelType*) to match a very wide variety of Pixel devices (ranging from pure-white LEDs to 5-color RGBWC LEDs)
        * **deprecates** constructor `Pixel(int pin, pixelType_t pixelType)` that limited *pixelType* to a pre-defined set of values.  This constructor will be removed in future versions of HomeSpan
      * adds new method `boolean Pixel::hasColor(char c)` to determine whether an existing Pixel object supports any particular color
        * **deprecates** `boolean Pixel::isRGBW()` which only distinguished RGB from RGBW devices.  This method will be removed in future versions of HomeSpan
      * upgraded the *PixelTester* sketch to interactively step users through selecting and testing the correct *pixelType* for their device
      * added new *Pixel-RGBWC* example demonstrating how to implement an RGBWC Pixel light-strip with separate Home App controls for the RGB and WC LED
      * see the **Pixels.md** page for full details

### ❗ IMPORTANT SIZE CONSIDERATIONS ❗

* version 3 of the Arduino-ESP32 Board Manager has a much larger footprint than version 2
* the same HomeSpan sketch compiled under 1.9.1 will be approximately 200K larger when compiled under HomeSpan 2.0.0-rc.1
* as a result, HomeSpan sketches will no longer fit into the *Default* partition scheme which only allocates 1.3MB to an App partition 
* HomeSpan sketches must instead be compiled under a larger partition scheme (such as *Minimal SPIFFS*, which provides for 1.9MB partitions)
* **HOWEVER - Espressif does not support changing partition tables via OTA!**
  * changes to the partition table are *ignored* if a sketch is updated via OTA
  * you can only update the partition table on a device if you load the sketch via the Serial port when connected to a computer
  * this means that if you have remote device running a HomeSpan sketch using the *Default* partition scheme, you will *not* be able to migrate to HomeSpan 2.0.0 using OTA
    * you must first connect the device directly to your computer, change the partition scheme to *Minimal SPIFFS*,  and then update the sketch via the Serial port
   * once HomeSpan has been successfully updated via the Serial Port using the new partition scheme, OTA can be used for any further updates (provided you don't need to subsequently change the partition scheme again for some other reason). 
      

<hr /><em>This discussion was created from the release <a href='https://github.com/HomeSpan/HomeSpan/releases/tag/2.0.0-rc.1'>HomeSpan 2.0.0-rc.1</a>.</em>

**Comment by oleksiikutuzov** [comment]
Thanks for your efforts porting the library to the new version of the board manager! Already tested with my LED Strip controller — SK6812 RGBW pixels are performing much more stable in animations. Previously, I got some random blinks on them even in a simple rainbow animation, thought it was my hardware issue (might be still 😄)

**Comment by blackjw1212** [comment]
Update HomeSpan 2.0.0-rc.1 WiFi.macAddress could not be obtained
display
*** HomeSpan Device ID and Pairing Data DELETED!  Restarting...

E (370544) wifi:NAN WiFi stop

**Reply by HomeSpan** [reply]
Sorry, I don't understand what you are trying to do or what the error was.  Please provide details.

**Reply by blackjw1212** [reply]
Thank you. You need to add
 #include <WiFi.h>
...
void setup() {
    Serial.begin(115200); 
    WiFi.begin();
...
  in this version. The display 00 problem has been corrected.

**Comment by MaffooClock** [comment]
For anyone else using PlatformIO, if you're getting this error:
```text
cannot open source file "FATAL_ERROR" (dependency of "HomeSpan.h")
```
The actual problem is with `src/extras/Pixel.h` and `src/extras/RFControl.h` both wanting `driver/rmt_tx.h` from IDF 5.  This was confusing for me since all of my PlatformIO dependencies were up-to-date.  Obviously, "up to date" doesn't mean "latest version."

I finally found the problem.  My `platformio.ini` had:
```ini
platform = platformio/espressif32
```
...but [the latest](https://github.com/platformio/platform-espressif32/releases/tag/v6.9.0) with Ardiuino framework is v2.0.17, which still uses IDF v4.

To get IDF 5 with the Arduino framework, I had to change my `platformio.ini` to:
```ini
platform = https://github.com/pioarduino/platform-espressif32/releases/download/stable/platform-espressif32.zip
```

Unfortunately, compiling still fails:
```text
HomeSpan/src/Network.cpp:30:10: fatal error: DNSServer.h: No such file or directory
```

**Reply by MaffooClock** [reply]
I just now found the discussion in PR #961 and learned that adding `lib_ldf_mode = deep` to platformio.ini resolved the problem with `DNSServer.h` not found.

Compiling succeeded, except at the very end when it complained that the compiled size exceeded the available flash size on my ESP-WROOM-32E board (using `board = esp32dev`), which by default uses a 1.25MB app partition.  This can be overridden by adding `board_build.partitions = huge_app.csv` to platformio.ini (in this case, the app partition was increased to 3MB).

In the end, the changes made to my `platformio.ini` file are now:
```ini
platform = https://github.com/pioarduino/platform-espressif32/releases/download/stable/platform-espressif32.zip
lib_ldf_mode = deep
board_build.partitions = huge_app.csv
```
... and compiling succeeds 🎉

**Reply by frankonski** [reply]
I use `board_build.partitions = min_spiffs.csv` instead of `huge_app.csv`.
And instead of using pioarduino (which is a fork of PIO), I stay with PIO but use real espressif libs this way:

```
platform = espressif32
board = esp32dev
framework = arduino
platform_packages=
  platformio/framework-arduinoespressif32 @ https://github.com/espressif/arduino-esp32.git#3.0.7
  platformio/framework-arduinoespressif32-libs @ https://github.com/espressif/esp32-arduino-lib-builder/releases/download/idf-release_v5.1/esp32-arduino-libs-idf-release_v5.1-632e0c2a.zip
lib_ldf_mode = deep
```

**Reply by MaffooClock** [reply]
> I use board_build.partitions = min_spiffs.csv instead of huge_app.csv.
> And instead of using pioarduino (which is a fork of PIO), I stay with PIO but use real espressif libs this way:

Aha, even better.  Thanks!
